### ISSUE TITLE
Create "Constant Raster" image size is different compared to "Rasterize"

### ISSUE THREAD

**Issue by aydin-gis** [qgis/QGIS#43547]
The Create Constant Raster Layer tool produces a different output size than the Rasterize (vector to raster) when using the same extent.

Tested in 3.16 and 3.18.
Steps to reproduce
Generate constant raster using the @map_extent as the desired extent (can do this in the model designer)
Rasterize a vector using the same pixel sizes, and the same @map_extent.

Expected results: Two rasters with identical height and width.
Actual result: two different sized rasters
Constant Raster:
Width	96
Height	51
Vector to Raster
Width	95
Height	50

Display.  The pink is the additional size of the constant raster
![image](https://user-images.githubusercontent.com/79950903/120726296-d69fe200-c52b-11eb-96cb-737f912eb99d.png)

Properties of Constant Raster
![image](https://user-images.githubusercontent.com/79950903/120726461-38f8e280-c52c-11eb-915d-2a78ed0666eb.png)
Properties of Rasterized Vector
![image](https://user-images.githubusercontent.com/79950903/120726433-28e10300-c52c-11eb-95f8-6592c7bba89a.png)

**Comment by elpaso** [comment]
Which one is the right output size?
Can you attach sample data to reproduce the issue?

**Comment by aydin-gis** [comment]
The smaller one is the correct one (from memory).
Changing the map extent, zoom, and window size has resulted in one being
off by a pixel in height, or width, or both height and width (as in my
screenshot).

Can be recreated by creating some basic lines, then passing the vector to
raster tool.
I can post sample data next week if you need.

My work around was to first generate my raster from vector using
@map_extent, then generating a constant raster based on the extent of the
rasterized vector.

**Comment by elpaso** [comment]
> The smaller one is the correct one (from memory). Changing the map extent, zoom, and window size has resulted in one being off by a pixel in height, or width, or both height and width (as in my screenshot). Can be recreated by creating some basic lines, then passing the vector to raster tool. I can post sample data next week if you need.

Yes, please. Try to leave out the @map_extent variable and use direct input values to the algorithm so that we can reproduce the issue easily, there is no need to use the modeler either: you should be able to reproduce the issue just using the algorithms directly.

I believe that there is a rounding difference in how the size is calculated but I need real values and data to reproduce the issue and confirm my hypothesis.

**Comment by agiudiceandrea** [comment]
@elpaso I can reproduce the issue on Windows using the processing tools with the following relevant parameters:

gdal:rasterize
INPUT: any vector layer (e.g. an empty memory layer) EPSG:32633
UNITS: Georeferenced units
WIDTH: 10
HEIGHT: 10
EXTENT: 656000,656184,4551000,4551184 [EPSG:32633]

native:createconstantrasterlayer
EXTENT: 656000,656184,4551000,4551184 [EPSG:32633]
TARGET_CRS: EPSG:32633
PIXEL_SIZE: 10

The output raster layers have the following dimensions in pixels and extent:

tool | width | height | extent
--- | --- | --- | ---
gdal:rasterize | 18 | 18 | 656000, 656180, 4551004, 4551184
native:createconstantrasterlayer | 19 | 19 | 656000, 656190, 4550994, 4551184

With EXTENT: 656000,656185,4551000,4551185 [EPSG:32633], the output raster layers have the following dimensions in pixels and extent:
tool | width | height | extent
--- | --- | --- | ---
gdal:rasterize | 19 | 19 | 656000, 656190, 4550995, 4551185
native:createconstantrasterlayer | 19 | 19 | 656000, 656190, 4550995, 4551185


It seems to me the different behaviour is due the different way GDALRasterize and QgsConstantRasterAlgorithm calculate width and height:
https://github.com/qgis/QGIS/blob/7f69678697ca65186682bfeb34beb3d72e07b526/src/analysis/processing/qgsalgorithmconstantraster.cpp#L154-L155
https://github.com/OSGeo/gdal/blob/master/gdal/apps/gdal_rasterize_lib.cpp#L505-L508
```c++
    nXSize = static_cast<int>(0.5 + (sEnvelop.MaxX - sEnvelop.MinX) / dfXRes);
    nYSize = static_cast<int>(0.5 + (sEnvelop.MaxY - sEnvelop.MinY) / dfYRes);
```
So, in QgsConstantRasterAlgorithm, width and height are always 'rounded up', while in GDALRasterize they are rounded to the nearest integer.

**Comment by aydin-gis** [comment]
I think agiudiceandrea was able to provide the feedback needed.
Reckon the feedback tag can be removed.
